tracing-subscriber/src/registry/sharded.rs

```diff
@@ -0,0 +1,297 @@
+    registry::{LookupSpan, SpanData},
+};
+use serde_json::{json, Value};
+use std::{
+    cell::RefCell,
+    collections::HashMap,
+    convert::TryInto,
+    fmt,
+    sync::{
+        atomic::{fence, AtomicUsize, Ordering},
+        Arc, Mutex,
+    },
+};
+use tracing_core::{
+    dispatcher,
+    field::{FieldSet, Visit},
+    span::{self, Id},
+    Dispatch, Event, Field, Interest, Metadata, Subscriber,
+};
+
+#[derive(Debug)]
+pub struct Registry {
+    spans: Arc<Slab<Data>>,
+}
+
+thread_local! {
+    static CONTEXT: RefCell<SpanStack> = RefCell::new(SpanStack::new());
+}
+
+#[derive(Debug)]
+pub struct Data {
+    metadata: &'static Metadata<'static>,
+    parent: Option<Id>,
+    children: Vec<Id>,
+    ref_count: AtomicUsize,
+    // TODO(david): get rid of these
+    values: Mutex<HashMap<&'static str, Value>>,
+}
+
+impl Drop for Data {
+    fn drop(&mut self) {
+        // We have to actually unpack the option inside the `get_default`
+        // closure, since it is a `FnMut`, but testing that there _is_ a value
+        // here lets us avoid the thread-local access if we don't need the
+        // dispatcher at all.
+        if self.parent.is_some() {
+            dispatcher::get_default(|subscriber| {
+                if let Some(parent) = self.parent.take() {
+                    let _ = subscriber.try_close(parent);
+                }
+            })
+        }
+    }
+}
+
+// === impl Registry ===
+
+impl Default for Registry {
+    fn default() -> Self {
+        Self {
+            spans: Arc::new(Slab::new()),
+        }
+    }
+}
+
+struct RegistryVisitor<'a>(&'a mut HashMap<&'static str, Value>);
+
+impl<'a> Visit for RegistryVisitor<'a> {
+    // TODO: special visitors for various formats that honeycomb.io supports
+    fn record_debug(&mut self, field: &Field, value: &dyn fmt::Debug) {
+        let s = format!("{:?}", value);
+        self.0.insert(field.name(), json!(s));
+    }
+}
+
+#[inline]
+fn idx_to_id(idx: usize) -> Id {
+    Id::from_u64(idx as u64 + 1)
+}
+
+#[inline]
+fn id_to_idx(id: &Id) -> usize {
+    id.into_u64() as usize - 1
+}
+
+impl Registry {
+    fn insert(&self, s: Data) -> Option<usize> {
+        self.spans.insert(s)
+    }
+
+    fn get(&self, id: &Id) -> Option<Guard<'_, Data>> {
+        self.spans.get(id_to_idx(id))
+    }
+
+    /// Decrements the reference count of the span with the given `id`, and
+    /// removes the span if it is zero.
+    ///
+    /// The allocated span slot will be reused when a new span is created.
+    fn drop_span(&self, id: Id) -> bool {
+        if !self.drop_ref(&id) {
+            return false;
+        }
+
+        // Synchronize only if we are actually removing the span (stolen
+        // from std::Arc);
+        fence(Ordering::Acquire);
+        self.spans.remove(id_to_idx(&id))
+    }
+
+    fn drop_ref(&self, id: &Id) -> bool {
+        let idx = id_to_idx(id);
+        let span = self.spans.get(idx).expect("Span not found");
```

Comment:
Nit/TIOLI: why not just use the `Registry::get` method which does the id/index covnersion?
```rust
let span = self.get(id).expect("Span not found");
```

Reply:
Also, since this can be called in a drop impl, it needs to be panic safe. So, we can't `expect` here. We should either check if we're already panicking before panicking here, or just always return `false` if the span is gone...

Reply:
did the former.